### PR TITLE
Fix #9591: Update station docking tiles upon placing a water object on a docking tile

### LINKED ISSUE #9591: [Bug]: Building a water object on a plain water docking tile does not correctly remove the docking tile
### Version of OpenTTD

master

### Expected result

Station::docking_station is always correct.

### Actual result

Building a water object on a plain water docking tile does not correctly remove the docking tile,
in that the station(s) associated with the docking tile do not have their docking_station tile areas updated.
This can result in multiplayer desyncs. This issue is detected by the cache checks in CheckCaches.

This was noticed because of griefers placing fake water tiles around docks, which later resulted in docking_station mismatches being logged.

### Steps to reproduce

* Set the desync log level to at least 2.
* Build a dock.
* Build water objects (e.g. from decoration NewGRF such as AuzWaterAndLandscapeObjects) on the dock's docking tiles.
* Note that a desync log message will be eventually logged.

### PR DESCRIPTION
**Pull request by SamuXarick**
## Motivation / Problem
Placing a water object on a docking tile doesn't run `CMD_LANDSCAPE_CLEAR` which calls `ClearTile_Water` to do the removal of docking tiles.

## Description
Borrow some part of the code from `ClearTile_Water` that deals with the removal of docking tiles and use it during object construction.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Review by glx22 — COMMENTED** [review]
Seems correct, but I think a savegame bump is needed, and an update of https://github.com/OpenTTD/OpenTTD/blob/master/src/saveload/afterload.cpp#L3109 with the new version.

File: src/saveload/afterload.cpp [hunk 1]
```diff
@@ -3106,8 +3106,9 @@ bool AfterLoadGame()
 		}
 	}
 
-	if (IsSavegameVersionBefore(SLV_DOCK_DOCKINGTILES)) {
-		/* All tiles around docks may be docking tiles. */
+	if (IsSavegameVersionBefore(SLV_REPAIR_OBJECT_DOCKING_TILES)) {
+		/* Placing objects on docking tiles was not updating adjacent station's
+		 * docking tiles. */
```
Comment (LordAro) [inline src/saveload/afterload.cpp]: unnecessary line break

**Review by LordAro — DISMISSED** [review]

File: src/object_cmd.cpp [hunk 2]
```diff
@@ -121,7 +122,9 @@ void BuildObject(ObjectType type, TileIndex tile, CompanyID owner, Town *town, u
 			Company::Get(owner)->infrastructure.water++;
 			DirtyCompanyInfrastructureWindows(owner);
 		}
+		bool remove = IsDockingTile(t);
 		MakeObject(t, owner, o->index, wc, Random());
+		if (remove) RemoveDockingTile(t);
```
Comment (TrueBrain) [inline src/object_cmd.cpp:127]: I honestly don't know the answer, but why does this need a variable? Can't we just do `RemoveDockingTile` before `MakeObject`?

Comment (JGRennison) [reply src/object_cmd.cpp:127]: No. RemoveDockingTile examines the state of the tile.

Comment (TrueBrain) [reply src/object_cmd.cpp:127]: So `RemoveDockingTileIfIFeelLikeIt`, gotcha. Tnx :)

**Review by TrueBrain — APPROVED** [review]
@glx22 and @LordAro seems to think this is correct, so lets go!

**Pull request merged**